### ISSUE TITLE
[RFC] Make it possible to initialize structs with just the module name

### ISSUE THREAD

**Issue by eriksvedang** [carp-lang/Carp#753]
When defining a sumtype you get a bunch of constructors for the various cases, like this:
```clojure
(deftype Sum A B)

(Sum.A)
(Sum.B)
```

When defining a product type you have to use `init` instead:

```clojure
(deftype Prod [x Int])

(Prod.init 123)
```

My suggestion is to allow the following as an alternative syntax:

```clojure
(Prod 123)
```

Right now this gives an error about trying to "call" the module, so hopefully we can just insert the call to init there instead (I haven't really looked into how to do the actual implementation).

As an added bonus, this would make the printing and reading of product types symmetrical also, since right now their string representation lacks the `.init` part:

```clojure
(str (Prod 123))
;;; =>
(Prod 123)
```

**Comment by TimDeve** [comment]
Slightly unrelated but I'm wondering if we should ape Rust for the typedef, so you can declare Prod as: 
```clojure
(deftype Prod [Int])
```
Pair could just become:
```clojure
(deftype Prod [a b])
```

**Comment by hellerve** [comment]
@TimDeve would all members be unnamed then?

**Comment by TimDeve** [comment]
I was thinking it would behave the same as here (numbered fields):
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a70992ebc6734b79332e6205928fbfb7

**Comment by hellerve** [comment]
I don’t really enjoy that, because it forces me to match on the entire thing if I want to change/update one field, and that gets very unwieldy on big types. I enjoy being able to have named getters and setters to update individual fields, and I actually think that the unnamed field pattern in Rust and Haskell is an antipattern. Consider, for instance the `DateTime` datatype underlying my `time` library:

https://github.com/carpentry-org/time/blob/c68425f37e102be10979d1dcfd453eadc2c23141/time.carp#L19-L28

This would become:

```clojure
(deftype Datetime [
  Int,
  Int,
  Int,
  (Maybe Int),
  (Maybe Int),
  (Maybe Int),
  (Maybe Int),
  (Maybe Timezone),
])
```

Except for the timezone, which is which? We can hope that the struct is ordered and the highest unit comes first, but we could also have opted for a US date format (day before month) or something like it. It unnecessarily creates confusion.

In fact, in the original proposal for sumtypes [I explicitly asked for having names for all of the members](https://github.com/carp-lang/Carp/issues/315#issuecomment-443528524), if only for documentation purposes. Sadly this was ignored. The community was in favor back then that this was an anti-pattern, but we just got a little complacent or a little too used to it I think.

Still, I’d strongly vote against doing this, and in fact I’d argue that we should go in the other direction.

I want to make very clear that this is in no way meant as an offense, and I value your opinion; I just came to a different conclusion.
